system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.cpp

```diff
@@ -0,0 +1,271 @@
+    "The topic to subscribe to, for calculating topic statistics.");
+  collect_topic_name_ = declare_parameter(
+    topic_statistics_constants::kCollectStatsTopicName,
+    topic_statistics_constants::kDefaultCollectStatsTopicName,
+    collect_topic_desc);
+  validateStringParam(collect_topic_name_);
+
+  const auto publish_topic_desc = buildTopicParameterDescriptor(
+    "The topic to publish topic statistics to.");
+  publish_topic_name_ = declare_parameter(
+    topic_statistics_constants::kPublishStatsTopicName,
+    system_metrics_collector::collector_node_constants::kStatisticsTopicName,
+    publish_topic_desc);
+  validateStringParam(publish_topic_name_);
+
+  initializeClock(clock_);
+
+  initializeCollectors(statistics_collectors_);
+
+  auto callback = [this](typename T::SharedPtr received_message) {
+      for (const auto & collector : statistics_collectors_) {
+        collector->OnMessageReceived(
+          *received_message,
+          getCurrentTime(clock_));
+      }
+    };
+  subscription_ = create_subscription<T>(collect_topic_name_, 10, callback);
+}
+
+template<typename T>
+SubscriberTopicStatisticsNode<T>::~SubscriberTopicStatisticsNode()
+{
+  if (RCL_RET_OK != rcl_clock_fini(&clock_)) {
+    RCUTILS_LOG_WARN("Failed to finalize clock");
+  }
+}
+
+template<typename T>
+void SubscriberTopicStatisticsNode<T>::StartPublisher()
+{
+  rcpputils::check_true(publish_timer_ == nullptr);
+
+  if (publisher_ == nullptr) {
+    publisher_ = create_publisher<MetricsMessage>(publish_topic_name_, 10);
+  }
+
+  publisher_->on_activate();
+
+  RCLCPP_DEBUG(this->get_logger(), "SetupStart: creating publish_timer_");
+  publish_timer_ = this->create_wall_timer(
+    publish_period_, [this]() {
+      this->PublishStatisticMessage();
+      this->ClearCollectorMeasurements();
+      this->window_start_ = this->now();
+    });
+  window_start_ = now();
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_activate(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_DEBUG(this->get_logger(), "on_activate");
+  for (const auto & collector : statistics_collectors_) {
+    if (!collector->Start()) {
+      return CallbackReturn::ERROR;
+    }
+  }
+
+  StartPublisher();
+  return CallbackReturn::SUCCESS;
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_deactivate(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_DEBUG(this->get_logger(), "on_deactivate");
+  for (const auto & collector : statistics_collectors_) {
+    if (!collector->Stop()) {
+      return CallbackReturn::ERROR;
+    }
+  }
+
+  StopPublisher();
+  return CallbackReturn::SUCCESS;
+}
+
+template<typename T>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+SubscriberTopicStatisticsNode<T>::on_shutdown(const rclcpp_lifecycle::State & state)
```

Comment:
You'll need to do some more work in this method. See the [possible state transitions](https://design.ros2.org/articles/node_lifecycle.html). ```PeriodicMeasurementNode``` is a good example of handling these for deactivate and shutdown (and check how they handle return if any errors).

Reply:
Modified these functions with `PeriodicMeasurementNode` as a reference. There are some differences because `SubscriberTopicStatistics` isn't itself a `Collector`, but the behavior is similar.